Design for new build tools for ROS 2 package cross-compilation (#233)
ros2/design: articles/cross_compilation_build_tools_design.md

location to compile with C and C++ cross-compilers.

On top of that we propose using a **colcon plugin `cc-build`** to wrap the instructions on the cross-compilation tutorial into a single
build command. That plugin would define a new `colcon_core.verb` entry point pointing to a class `CCBuildVerb` that uses the
[Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
used by `colcon build` with additional functionality to handle the sysroot and toolchain file, and to inject additional `cmake-args` arguments.
That would allows us to run the single command below from a ROS 2 overlay workspace on a developer workstation, for cross-compiling
the packages in the workspace up to a package `performance_test` for the ARM-HF platform.

```bash
colcon cc-build --arch armhf --os ubuntu_bionic \
  --packages-up-to performance_test # other `colcon build` arguments
```

That would should result on a new workspace directory `armhf-ubuntu_bionic` with subdirectories `build`, `install`, and `log` directories, with the 

> ross-compiled binaries. All options for `colcon build` will be supported by forwarding them to the `BuildVerb` class. That would be the only command required for cross-compiling the packages in a workspace. Under the hood, the plugin would do the following:

Reviewer: ross-compiled -> cross-compiled

Author: I'll fix it, thanks

- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
 (`qemu-user-static`, `g++-aarch64-linux-gnu`, ..) should be installed in the development workstation used for the build. The `cc-build` plugin will
 declare those system dependencies using a `stdeb.cfg` file like is done on [other colcon plugins](https://github.com/colcon/colcon-core/blob/master/stdeb.cfg).

> - In order to build the sysroot, the plugin downloads a base ROS 2 Docker image for the target platform, and builds a workspace dependent sysroot image by doing a `COPY` to copy the contents of the workspace into the container, and running `rosdep` on the copy. That ensures no workspace system dependency is missing. Note we we can run `rosdep`

Reviewer: Double we in "Note we we can run"

Author: Thanks for letting me know about the typo, it's fixed in a later commit.

Regarding the list of supported platforms, after some feedback in this PR the idea is that instead of a colcon plugin we'd add a command for setting up the sysroot, and a colcon mixin for adding the required cross compilation flags to `colcon build`. For example we might build ROS 2 for armhf as follows

```bash
# Setup the sysroot for the platform (generic_armhf, ubuntu_bionic, ROS crystal, RMW FastRTPS)
# this leaves the sysroot at generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot
cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic --distro crystal --rmw fastrtps \
   --sysroot-base-image ros2:armhf_bionic_crystal_base-crystal
# this creates some symlinks like /lib/arm-linux-gnueabihf -> $(pwd)/generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot/lib/arm-linux-gnueabihf in line with the "Hack to find Poco" from https://github.com/ros2/cross_compile/blob/master/entry_point.sh#L38
bash generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_system_setup.bash

# this setups some env vars for cross compilation
source generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_build_setup.bash

# Run colcon build using the mixin for this platform
colcon build --mixin cc-generic_armhf-ubuntu_bionic-fastrtps-crystal
```

The image `ros2:armhf_bionic_crystal_base-crystal` above refers to a Docker image with just the basics depedencies to build ROS2. To cross compile a workspace with additional packages that are not part of ROS 2, we use a different argument for `--sysroot-base-image` that corresponds to an image with ROS 2 pre built for that architecture. Then `cc_build_setup.bash` would source `generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot/opt/ros/crystal/setup.bash` as part of the setup.

This means we'd have a different colcon mixin for each platform.

Reviewer 2: > This means we'd have a different colcon mixin for each platform.

If the duplication between them becomes an issue we could consider supporting one mixin to extend another mixin - just a thought.

> in Docker for the ARM-HF platform thanks to [qemu](https://www.qemu.org/). That image's file system is then exported by launching a container a running `docker container export` on it, to a path following a sensible convention, like a workspace-local `armhf-ubuntu_bionic/sysroot` directory, so it can be used as a sysroot during the build.

Reviewer: Extra "a" in "...launching a container a running..."

Author: I'll fix it, thanks

argument for `docker image pull`, and that can be used in `FROM` Dockerfile statement. If it is not specified, the plugin will deduce the value of
`--sysroot-base-image` from the `--arch` and `--os` arguments, for the architecture-OS combinations that have an image published for them on
[osrf/ros2 on Docker Hub](https://hub.docker.com/r/osrf/ros2/).
- `--sysroot-image`: Specifies a Docker image for the workspace sysroot. The build will export the sysroot from that image, instead of creating a sysroot image from
the workspace. No check is performed to ensure the image has all the dependencies required by the packages in the workspace.
- `--force-sysroot-build`: Boolean flag to force exporting the sysroot even if it already available in the workspace. This is useful for example when new system
dependencies are added to a package in the workspace. This leads to rebuilding the dependent sysroot image after the `COPY` statements that copies the workspace into
the image, including running `rosdep install` from scratch. This does not rebuild the base ROS 2 Docker image.

> - `--sysroot-path`: Specifies a local path for an existing sysroot. This can be useful for example if different workspaces share the same set of system dependencies.

Reviewer: --sysroot-path is also very useful to target products handling their file system outside Docker. In other words, the whole built-in support for preparing a sysroot being proposed under this design is a great way for a developer to focus on the "arch" support, but in practice, companies and individuals might be using other options to create their file system: Buildroot, Yocto, other pre-built distros, etc.

Author: With the new design with a separate `cc-setup-sysroot` command we could add a `--sysroot-path` optional argument to it, so  `cc-setup-sysroot`  would just generate the `cc_system_setup.bash` and `cc_build_setup.bash` when that argument is present. That should work provided the sysroot contains standard Linux paths line `/lib` and `/usr/lib/`

- `--use-base-image-sysroot`: Boolean flag, when true the base ROS 2 Docker image is used to build the sysroot, so the dependencies declared in packages in the current
workspace are ignored. This is useful to speed up builds for packages without additional dependencies.

Support for other platforms can be added by creating corresponding ROS 2 Docker images, and variants of the toolchain file if required.

> We could include support for ARM-HF and ARM64 in the first release of the plugin.

Reviewer: You may also to restrict the valid OS (when using --os) to Ubuntu to get started.

Author: Thanks for the suggestion, other comments in this PR had similar concerns. After incorporating that feedback that idea now is that the platform is determined by a tuple ` ('arch', 'os', 'distro', 'rmw')`, for example `(generic_armhf, ubuntu_bionic, ROS crystal, RMW FastRTPS)`, as seen in my previous comment. That would allow extending this in the future with support for more specific ECUs than `generic_armhf`, as suggested by @dejanpan above

file for the platform.

We could then test cross-compile support for the new platform by building the base  ROS 2 Docker image, and a custom sysroot image for some reference package
(e.g. [ApexAI/performance_test](https://github.com/ApexAI/performance_test/)), and then launching a Docker container for the that sysroot image, running
`colcon test` using the cross-compiled binaries. The container would be launched with a [bind mount](https://docs.docker.com/storage/bind-mounts/) of the
workspace so they have access to the compiled binaries.

In order to simplify the development of base ROS 2 Docker images for new architecture-OS combinations, on a second iteration of this plugin we might add a
new verb `colcon cc-build-sysroot-base-image` that will 1) copy the required qemu files and ROS 2 source into a temporary file; 2) build a Docker image from a
specified Dockerfile, with access to the those resources; 3) optionally publish the image into a Docker registry, using a suitable naming convention like
`${docker_repo_uri}:${os}_${arch}_${prebuilt?}_${rosdistro}`. Dockerfiles compatible with this command, for each architecture-OS pair would be keep on the

> `ros2/cross_compile` repository, so they can be used in CI/CI pipeline if needed.

Reviewer: CI/CI pipeline -> CI/C**D** pipeline ?

Author: I'll fix that, thanks

Reviewer 3: I don't know about you but I feel the need to do CI on my CI all the time. Who watches the watchers and all that.

---
layout: default
title: Cross-compiling ROS 2 packages
abstract:
  This article is a design proposal for extending the build tools for ROS 2 to

>   simplify cross-compiling ROS 2 packages for platforms currently not supported

Reviewer 4: I would suggest that we define what a platform is, e.g. in the automotive world a platform is a tupple of {ECU:OS:Framework}. So for instance an example of a platform would be: {Renesas R-Car H3:QNX:ROS2 C}.
Note that Renesas R-Car H3 implicitly implies that this is an `aarch64` architecture for computer processor.

Since ROS 2 depends greatly on an underlying middle-ware it would be good to even include that into the tupple, so {ECU:OS:Middle-ware:Framework}.

Reviewer 3: Perhaps we can have a generic part and a platform-specific set of samples? Or a small set of supported platforms, complying with the relevant REP?

Author: I agree with the proposal to add the ROS distro and rmw implementation. For the ECU I propose starting with a `generic_armhf`, and then adding some specific ECUs as @gbiggs suggests. So we for example we could have `generic_armhf-ubuntu_bionic-fastrtps-crystal` as an example  platform identifier.

  by ROS 2, or building on a host with different platforms than the target

>   platform. That later use case is particularly useful for speeding up build

Reviewer 5: "latter" is the proper spelling here.

Reviewer 6: Posing @esteve's remark as a suggestion.

```suggestion
  platform. That latter use case is particularly useful for speeding up build
```

  times by building the code on a powerful development workstation, and then
  deploying the binaries to a constrained resources hardware that uses the
  target platform, thus making the development cycle much faster.
author: >
  [Thomas Moulard](https://github.com/thomas-moulard),
  [Juan Rodriguez Hortala](https://github.com/juanrh)
published: false
---

{:toc}

# {{ page.title }}

<div class="abstract" markdown="1">
{{ page.abstract }}
</div>

Original Author: {{ page.author }}

## Background

> The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)

Reviewer 5: Just a minor note, both https://github.com/esteve/ros2_java and https://github.com/esteve/ros2_objc predate the instructions that ARM wrote up. https://discourse.ros.org/t/crosscompiling-ros2-for-raspberry-pi-or-pretty-much-any-other-arm-based-board/2995 used the same technique as ARM's instructions (Docker + QEmu), but for armhf as well.

Author: I'll add that precedents to the document, thanks for bringing them to our attention

contains instructions for cross-compiling ROS 2 to unsupported platforms, using
some configuration files for CMake and Docker that is setting maintained in the
[ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
However, that tutorial doesn't provide instructions to build other ROS 2

> packages, but just for building a whole ROS 2 distribution. Ideally, we should

Reviewer 7: The "Cross-compiling against a pre-built ROS2" section aims to provide an example on how to CC a single ros2 package without building the full distribution.

Author: I'll remove that statement, and make it clear that this proposal only contributes build tools to make it easier to use the existing instructions

be able to cross-compile a ROS package for ROS 2 by launching a single
`colcon build` command. Finally, cross-compiling for other platforms besides
ARM64 is not supported in the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).

> 

Reviewer 5: Worth mentioning is https://github.com/ruslo/polly, which contains a bunch of toolchains for crosscompiling CMake projects for a variety of platforms.

Reviewer 8: Sounds awesome, thanks!

Reviewer 7: armhf is also supported with the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/). The architecture is an argument of the "sysroot dockerfile"

> Use an argument to specify the architecture of the target-file system.
This can be specified with the `--build-arg` argument on the docker
build command:
  *for armhf: docker build --build-arg ARM_ARCH=arm32v7 ...
  *for aarch64: docker build --build-arg ARM_ARCH=arm64v8 ...

Author: [Dockerfile_ubuntu_arm](https://github.com/ros2/cross_compile/blob/master/sysroot/Dockerfile_ubuntu_arm) indeed supports that argument, but [entry_point.sh](https://github.com/ros2/cross_compile/blob/master/entry_point.sh#L28) assumes ARM64 when it sets `export TARGET_TRIPLE=aarch64-linux-gnu`. However I agree we could be more precise and state that the tutorial does support ARMHF, but the automation described in the "Automated Cross-compilation" section does not fully support it. I'll be more precise about that, as this proposal doesn't try to change the basic approach to cross compilation, but to pack it into a development tool that is easy to use

Reviewer: The Dockerfiles and scripts under ros2/cross_compile are just exampless to go along with the tutorial explaining cross-compilation and what it takes to achieve it (independent of the arch and target platform). The idea of integrating this support into Colcon is great. I assume the proposal here is to have a list of supported platforms (e.g. armhf) instead of just deriving the arch/plat from the new Colcon plugins, right?

## Proposed Approach

We propose continuing with the general approach of the cross-compilation
tutorial, based on building a sysroot for the target platform using qemu and
Docker, and then using `cmake-toolchains` with a toolchain file pointing
`CMAKE_SYSROOT` to the sysroot location to compile with C and C++
cross-compilers.

> On top of that we propose using a **colcon plugin `cc-build`** to wrap the

Reviewer 5: Instead of adding a new verb, wouldn't it more appropriate to add this logic to the CMake extension for colcon? `cc-build` seems to assume that all packages that are to be crosscompiled use CMake.

Reviewer 8: The tricky part is the sysroot management which is based on Docker. I think it's too brutal to directly pull a dependency on Docker in all cases when one is using `colcon`. Being in a plugin is a convenient way to get users to "opt-in".

Reviewer 3: If it's added to the CMake extension, then wouldn't that cause it to exclude packages that only use Python? I know that they don't need cross-compiling, but it would be convenient if `colcon` could build an entire application in the crosscompile target file system, not just the parts that are done with CMake, so that it is really easy for the user to just copy the install directory to the target device.

Reviewer 5: In addition to CMake and Python, `colcon` supports Gradle, Bazel, Cargo and probably more, hence why I believe this functionality should be part of each build plugin, because it's hard to cover them all in a single plugin. I'm thinking something like `colcon build --cmake-sysroot ... --python-sysroot ...` would work. Perhaps we could enforce that any new build tool that gets integrated into `colcon` has support for cross compiling.

Reviewer 2: > Perhaps we could enforce that any new build tool that gets integrated into `colcon` has support for cross compiling.

This goes against one of the [stated software design goals](https://colcon.readthedocs.io/en/released/developer/design.html#software-design) of `colcon`:

    > Each functionality added should follow the principle “you don’t pay for what you don’t use”.

So if a plugin proving support for X doesn't need sysroot support I shouldn't need to implement it.

Reviewer 9: Related to this see below where I suggest having the platform support builds embedded into docker images instead of on the host. So that you don't need to have the cross compilation resources on your local machine, just a docker dependency.

instructions on the cross-compilation tutorial into a single build command.
That plugin would define a new `colcon_core.verb` entry point pointing to a
class `CCBuildVerb` that uses the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern)
to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
used by `colcon build` with additional functionality to handle the sysroot and
toolchain file, and to inject additional `cmake-args` arguments. That would
allows us to run the single command below from a ROS 2 overlay workspace on a
developer workstation, for cross-compiling the packages in the workspace up to
a package `performance_test` for the ARM-HF platform.

```bash
colcon cc-build --arch armhf --os ubuntu_bionic \
  --packages-up-to performance_test # other `colcon build` arguments
```

> That would should result on a new workspace directory `armhf-ubuntu_bionic`

Reviewer 4: remove `should`

Reviewer 4: on => in

Author: Fixed

with subdirectories `build`, `install`, and `log` directories, with the

> ross-compiled binaries. All options for `colcon build` will be supported by

Reviewer 4: ross => cross

I actually propose to proof read this PR for spelling mistakes and other grammar.

Author: I have made a pass to fix spelling and grammar mistake, please let me know if you find something else

forwarding them to the `BuildVerb` class. That would be the only command
required for cross-compiling the packages in a workspace. Under the hood, the
plugin would do the following:

- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
  (`qemu-user-static`, `g++-aarch64-linux-gnu`, ...) should be installed in the
  development workstation used for the build. The `cc-build` plugin will
  declare those system dependencies using a `stdeb.cfg` file like is done on
  [other colcon plugins](https://github.com/colcon/colcon-core/blob/master/stdeb.cfg).
- In order to build the sysroot, the plugin downloads a base ROS 2 Docker image
  for the target platform, and builds a workspace dependent sysroot image by
  doing a `COPY` to copy the contents of the workspace into the container, and

>   running `rosdep` on the copy. That ensures no workspace system dependency is

Reviewer 8: I'm wondering whether we could execute a "well-known script" like `./initialize.sh` or something to avoid making the assumption that `rosdep` will always be the only required step there.

This would be helpful to test the plugin too as we could run non-ROS 2 images (e.g. base OS image only) to make sure we can cross-compile a simple "hello world" binary.

Author: That makes sense, but I'm not sure where to put it. Some options that come to my mind: 

- Make it dynamic: and optional argument for the plugin to provide and `./initialize.sh` to run at the end of the workspace image. If not provided we use a default `./initialize.sh` that runs rosdep. 
- Make it static: for each (os, arch) combination (or whichever definition of platform we use) the plugin uses a different Dockerfile,  extending a base Dockerfile that extends the base image for ROS 2 specified and does the COPY of the workspace contents. This means the plugin has to be modified to add or modify the Dockerfile specific for a platform. 

Does some of these option sound good to you?

Reviewer 9: I would suggest holding off on this until it's necessary for a use case instead of speculatively extending the scope.

In order to simplify the development of base ROS 2 Docker images for new
architecture-OS combinations, on a second iteration of this plugin we might add
a new verb `colcon cc-build-sysroot-base-image` that will 1) copy the required
qemu files and ROS 2 source into a temporary file; 2) build a Docker image from
a specified Dockerfile, with access to the those resources; 3) optionally
publish the image into a Docker registry, using a suitable naming convention like
`${docker_repo_uri}:${os}_${arch}_${prebuilt?}_${rosdistro}`. Dockerfiles
compatible with this command, for each architecture-OS pair would be keep on the
`ros2/cross_compile` repository, so they can be used in CI/CI pipeline if needed.

### Advantages of the approach

The proposed design leads to a simple development experience, where cross
compilation is triggered by a one-liner command.
The design is extensible and new platforms can be supported easily.

For simple packages with additional dependencies, there is no need to build the
sysroot locally with Docker, and a readily available base Docker image can be
used. For packages with custom dependencies, the base sysroot can be extended

> using `rosdep install`.

Reviewer 5: I'd clarify that `rosdep install` here needs to be run inside QEmu.

Reviewer 8: (through `binfmt_misc`)

Author: I'll add explicit reference to the QEmu usage there

used by `colcon build` with additional functionality to handle the sysroot and
toolchain file, and to inject additional `cmake-args` arguments. That would
allows us to run the single command below from a ROS 2 overlay workspace on a
developer workstation, for cross-compiling the packages in the workspace up to
a package `performance_test` for the ARM-HF platform.

```bash
colcon cc-build --arch armhf --os ubuntu_bionic \
  --packages-up-to performance_test # other `colcon build` arguments
```

That would should result on a new workspace directory `armhf-ubuntu_bionic`
with subdirectories `build`, `install`, and `log` directories, with the
ross-compiled binaries. All options for `colcon build` will be supported by
forwarding them to the `BuildVerb` class. That would be the only command
required for cross-compiling the packages in a workspace. Under the hood, the
plugin would do the following:

- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
  (`qemu-user-static`, `g++-aarch64-linux-gnu`, ...) should be installed in the

>   development workstation used for the build. The `cc-build` plugin will

Reviewer 9: I would like to suggest a different approach that will not require the installation of the cross compile toolchain and development tool on the users development machine. Instead we can create a docker image which contains the toolchain and all the cross compilation tools. This allows us to ensure that the build is much more reproducible by allowing us to validate the versions of the development tools in the image with the toolchain. And the only content coming from the users environment is the workspace.

We can still generate a sysroot that can be mounted in for testing and providing dependencies. But the sysroot can be  mounted or COPY'd into the built container using multistage docker builds. 

This will break the need for the plugin to have dependencies on the tools. This will be especially important as we start to try to support new architectures. Requiring all the tools to be installed locally would become cumbersome.

Similarly it allows us to QA the tools on a specific platform and can run reliably even if the user is not running on one of those platforms. Say the user is running on an older distro without the necessary recent enough version of the gcc armhf, if the tools are up to date in the bionic image it can build fine on their machine.

Reviewer 2: While doing the build within Docker sounds like a good idea I think the (to be developed?) tools should be not limited to that use case. Imo there should be "tools" which can do they job (in whatever environment) and then there can be a specific "meta tool" which uses the tools in a certain way - e.g. within a Docker container. Developing something only for a very use case will likely make the result not really reusable and be only useful to a small subset of users.

Author: Another proposal would be splitting this into two commands, as suggested in this thread several times:

- A new command `setup-sysroot` would use Docker to generate the sysroot
- A new colcon mixin that adds options for using the sysroot to the colcon build task. This command requires having the cross compilation toolchain installed.
- On an second iteration we might:
  - Maintain a docker image for cross compilation that has installed both the cross compilation toolchain, and the aforementioned colcon mixin, based on the images at the ros2/cross_compile repository.
  - Implement a new `cross-build` command (definitive name TBD) that will ensure the sysroot is created using `setup-sysroot`, and then launch the cross compilation into the docker container with the cross compilation toolchain.

How does that sound?

Reviewer 2: > A new command `setup-sysroot` would use Docker to generate the sysroot

What exactly is this *verb* supposed to be doing? (Depending on the answer) why should this be a `colcon` verb?

> A new colcon mixin that adds options for using the sysroot to the colcon build task. This command requires having the cross compilation toolchain installed.

Sounds good to me.

Author: > What exactly is this verb supposed to be doing? (Depending on the answer) why should this be a colcon verb?

The idea is keeping a directory in the workspace for that target platform, for example  `armhf-ubuntu_bionic`, determined by convention based on arguments for architecture and OS. Let us call that directory the cc-root. The command `setup-sysroot` would create a `sysroot` subdirectory of the cc-root (e.g.  `armhf-ubuntu_bionic/sysroot`) with the sysroot for the target platform. The colcon mixin for cross-compilation would then use the same convention as `setup-sysroot` to determine the cc-root, and use the sysroot stored there. We'd have mixins that provide shortcuts for cross compiling known architecture-os combinations. 

Here I'm not proposing having `setup-sysroot` as a new colcon verb, but as a new separate command. I'm also open to use a different name instead of "setup-sysroot".

Reviewer 2: > Here I'm not proposing having setup-sysroot as a new colcon verb, but as a new separate command.

Oh, outside of `colcon`? Please feel free to choose whatever you like - I don't have an opinion on that.

Author: I have updated the design to replace the colcon plugin by the new command and the colcon mixin, please let me know what you think

Reviewer 2: Sounds reasonable to me. With this update I might not be the best person to provide more feedback since the `colcon` part of this boils down to a mixin. Maybe it would be good to update the title of the ticket to reflect the current state?

I am not sure I understand how the user gets the CMake toolchain file which would be referenced from the mixin. But that might be a too detailed question for now.

contain instructions for cross-compiling ROS 2 to unsupported platforms like
ARM64 and ARM-HF, using some configuration files for CMake and Docker.
Other projects like [steve/ros2_raspbian_tools](https://github.com/esteve/ros2_raspbian_tools),
[esteve/ros2_java](https://github.com/esteve/ros2_java) provide support
for additional platforms like Android and iOS.
Ideally, we should be able to cross-compile a ROS package for ROS 2 by
launching a single command.
That command should also be implemented using an extensible
design that allows supporting new platforms with low effort.

## Proposed Approach

We propose continuing with the general approach of the cross-compilation
tutorial, based on building a sysroot for the target platform using QEMU and
Docker, and then using [`cmake-toolchains`](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html)
with a toolchain file pointing `CMAKE_SYSROOT` to the sysroot location, to
compile with C and C++ cross-compilers.

We propose adding a set of commands to wrap the instructions on the
cross-compilation tutorial.
Those commands would use a new workspace

> directory for the target platform, determined by a convention based on

Reviewer 10: typo: "on on"

on the target ECU, operating system, rmw implementation, and ROS distribution.
Let us call that directory the _cc-root_.
Under that convention, the ECU
implies the architecture, but generic ECUs like `generic_armhf` would also be
available. For example `generic_armhf-ubuntu_bionic-fastrtps-crystal` is an
example _platform identifier_.
Specifically, we would add the following commands:

- A new `cc-setup-sysroot` command would use Docker to generate the sysroot
  for the target platform, and store it in a `sysroot` subdirectory of the
  cc-root.
  This command would take arguments to specify the target platform,
  for example:

```bash
cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic \
                 --rmw fastrtps --distro crystal
```

- A new colcon mixin for each known platform, which adds options to the colcon

>   build task for using asysroot generated using `cc-setup-sysroot`, by using

Reviewer 10: ```suggestion
  build task for using a sysroot generated with `cc-setup-sysroot`, by using
```

  the same path conventions.

>   For example from a ROS 2 overlay workspace on a

Reviewer 10: ```suggestion
  For example, from a ROS 2 overlay workspace on a
```

  developer workstation, the following command would cross-compile the packages
  in the workspace up to a package `performance_test` for the platform
  specified.
  Note this command requires having the cross [compilation
  toolchain](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
  installed.
  That would create `build`, `install`, and `log`
  subdirectories of the cc-root `generic_armhf-ubuntu_bionic-fastrtps-crystal`,
  with the cross-compiled binaries.

```bash
colcon build --mixin cc-generic_armhf-ubuntu_bionic-fastrtps-crystal \
  --packages-up-to performance_test # any other other `colcon build` arguments
```

- On an second iteration of this initiative we would:
  - Maintain a docker image for cross compilation, that has both the cross
  compilation toolchain and the aforementioned colcon mixins installed.
  - Implement a new `cc-build` command that would ensure the sysroot is created
  using `cc-setup-sysroot`, and then launch the cross compilation into a docker
  container using the corresponding Docker image.

Under the hood, the `cc-setup-sysroot` command would do the following:

> - The command downloads a base ROS 2 Docker image for the target platform, and

Reviewer 10: Consider breaking this long point into a numbered list for clarity. E.g.

> 1. Downloads a base ROS 2 Docker image for the target platform.
> 2. Builds a workspace dependent sysroot image by running `COPY` to get the contents of the workspace into the container.
> 3. Run `rosdep` to ensure no workspace system dependency is missing.
> 4. etc...

  builds a workspace dependent sysroot image by
  running `COPY` to get the contents of the workspace into the container, and
  running `rosdep` on the copy.
  That ensures no workspace system dependency is
  missing.

>   Note we can run `rosdep` in Docker for platform like ARM-HF thanks

Reviewer 10: ```suggestion
  Note we can run `rosdep` in Docker for platforms like ARM-HF thanks
```

  image from the workspace.
  No check is performed to ensure the image has all the
  dependencies required by the packages in the workspace.
- `--use-base-image-sysroot`: Boolean flag, when true the base ROS 2 Docker
  image is used to build the sysroot, so the dependencies declared in packages in
  the current workspace are ignored.
  This is useful to speed up builds for
  packages without additional dependencies.

Support for other platforms can be added by creating corresponding ROS 2 Docker
images, and variants of the toolchain file if required.
We could include support for generic versions of ARM-HF and ARM64, on Ubuntu
Bionic with ROS 2 crystal and fastrtps, in the first release of the commands.

### Frequently asked questions

- _How can we run the resulting binaries?_ Packaging the resulting binaries is
  outside of the scope of this document.
  The cross-compiled binaries will be
  available in the workspace in the cc-root (e.g.
  `generic_armhf-ubuntu_bionic-fastrtps-crystal`).
  We make the assumption that
  the user will copy the whole workspace to the remote device (including `src`)
  and run `rosdep` on the target platform, to install the dependencies before

>   running the binaries.

Reviewer 2: Just a note: this will also install all build dependencies which would't be necessary.

### Maintenance and testing

Regarding maintenance, in order to support cross-compilation for a new platform
we would need to:

- Prepare and publish a base ROS 2 Docker image for the new platform, both
  fetching the dependencies using the ROS 2 source using `rosdep install`, and
  from the pre-built binary if available for that platform.
- Adapt the toolchain file for the new platform, if required.
  For that the toolchains in [ruslo/polly](https://github.com/ruslo/polly) might be useful.

> - Add a new `cross-compile*` colcon mixin to for the new platform, add perform

Reviewer 10: ```suggestion
- Add a new `cross-compile*` colcon mixin to for the new platform and perform
```

  any modifications needed to use a suitable toolchain file.

> We could then test cross-compile support for the new platform by building the

Reviewer 10: Consider making this a numbered list as well. E.g.

> We can then test cross-compile support for the new platform by:
> 1. Building the base ROS 2 Docker image.
> 2. Building a custom sysroot image for a reference package.
> 3. etc...

base ROS 2 Docker image, and a custom sysroot image for some reference package
(e.g. [ApexAI/performance_test](https://github.com/ApexAI/performance_test/)),
and then launching a Docker container for the that sysroot image, running
`colcon test` using the cross-compiled binaries.
The container would be launched with a [bind mount](https://docs.docker.com/storage/bind-mounts/)
of the workspace so they have access to the compiled binaries.

In order to simplify the development of base ROS 2 Docker images for new
architecture-OS combinations, on a future iteration of this initiative we might add

> a new command `build-sysroot-base-image` that will 1) copy the required

Reviewer 10: Consider making this a proper markdown list.

so they can be used in CI/CD pipeline if needed.

### Advantages of the approach

The proposed design leads to a simple development experience, where cross
compilation is triggered by two simple commands, or a single `cc-build` command
for the second iteration.
The design is extensible and new platforms can be supported easily.

For simple packages with additional dependencies, there is no need to build the
sysroot locally with Docker, and a readily available base Docker image can be
used.
For packages with custom dependencies, the base sysroot can be extended
using `rosdep install` on a Docker image that uses QEMU through [`binfmt_misc`](https://en.wikipedia.org/wiki/Binfmt_misc).

## Limitations and open questions

This setup is focused on cross-compiling for Linux based platforms.
For example we couldn't use this to cross-compile for Windows, because we cannot install
the Visual Studio compiler using Apt.
On the other hand, a Windows or Mac development workstation could use this
build setup, running the build commands inside a Docker container, although
that would imply some performance degradation due to the virtualization layer
that Docker employs on those platforms.

> We should determine a process for building an publishing the base ROS 2 Docker

Reviewer 10: ```suggestion
We should determine a process for building and publishing the base ROS 2 Docker
```

---
layout: default
title: Cross-compiling ROS 2 packages
abstract:
  This article is a design proposal for extending the build tools for ROS 2 to
  simplify cross-compiling ROS 2 packages for platforms currently not supported
  by ROS 2.
author: >
  [Thomas Moulard](https://github.com/thomas-moulard),
  [Juan Rodriguez Hortala](https://github.com/juanrh)
published: false
---

{:toc}

# {{ page.title }}

<div class="abstract" markdown="1">
{{ page.abstract }}
</div>

Original Author: {{ page.author }}

## Background

The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
and the [steve/ros2_raspbian_tools repository](https://github.com/esteve/ros2_raspbian_tools)
contain instructions for cross-compiling ROS 2 to unsupported platforms like
ARM64 and ARM-HF, using some configuration files for CMake and Docker.

> Other projects like [steve/ros2_raspbian_tools](https://github.com/esteve/ros2_raspbian_tools),

Reviewer 11: Looks like
`[steve/ros2_raspbian_tools](https://github.com/esteve/ros2_raspbian_tools)` is already mentioned above. Did you mean https://github.com/esteve/ros2_objc  (for iOS) ?

Docker, and then using [`cmake-toolchains`](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html)
with a toolchain file pointing `CMAKE_SYSROOT` to the sysroot location, to
compile with C and C++ cross-compilers.

We propose adding a set of commands to wrap the instructions on the
cross-compilation tutorial.
Those commands would use a new workspace
directory for the target platform, determined by a convention based on
the target ECU, operating system, rmw implementation, and ROS distribution.
Let us call that directory the _cc-root_.
Under that convention, the ECU
implies the architecture, but generic ECUs like `generic_armhf` would also be
available. For example `generic_armhf-ubuntu_bionic-fastrtps-crystal` is an
example _platform identifier_.
Specifically, we would add the following commands:

- A new `cc-setup-sysroot` command would use Docker to generate the sysroot
  for the target platform, and store it in a `sysroot` subdirectory of the
  cc-root.
  This command would take arguments to specify the target platform,
  for example:

```bash
cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic \

>                  --rmw fastrtps --distro crystal

Reviewer 6: ```suggestion
                 --rmw fastrtps --rosdistro crystal
```

--rosdistro, abbreviated `-r` would be more consistent with other tools in the ROS community and less likely to be misinterpreted as a Linux ("distro") distribution.

Reviewer 6: An rmw can consist of more than the underlying DDS mechanism. For example there are static and dynamic typesupport variants of rmw_fastrtps. On some level only the underlying middleware is significant for setting up the sysroot as it defines what's available for rmw implementation packages to be built on top of it, but I do want to make sure that we're not equating **one** middleware provider with **one** rmw implementation.

1. Downloads a base ROS 2 Docker image for the target platform.
1. Builds a workspace dependent sysroot image by using a Dockerfile that starts
   from that base image and:
   1. Runs `COPY` to get the contents of the workspace into the container.
   1. Uses `rosdep` to ensure no workspace system dependency is missing.
      Note we can run `rosdep` in Docker for platforms like ARM-HF thanks
      to [QEMU](https://www.qemu.org/).
1. Launches a container for that image and exports its file system into a
   `sysroot` subdirectory of the cc-root.
1. During a build, the colcon mixin sets the variable `SYSROOT` to the sysroot path,
   as well as any other variables required to configure the toolchain file [generic_linux.cmake](https://github.com/ros2/cross_compile/blob/master/cmake-toolchains/generic_linux.cmake).
   It also sets `DCMAKE_TOOLCHAIN_FILE` through `cmake-args`, so it points to that toolchain file.

The ROS 2 base images are variants of [sysroot/Dockerfile_ubuntu_arm](https://github.com/ros2/cross_compile/blob/master/sysroot/Dockerfile_ubuntu_arm)
from ros2/cross_compile.
OSRF would publish base ROS 2 images with ROS 2 prebuilt, that should be used

> for cros compiling workspaces with user packages.

Reviewer 6: ```suggestion
for cross compiling workspaces with user packages.
```
Just fixing spelling.

We could include support for generic versions of ARM-HF and ARM64, on Ubuntu
Bionic with ROS 2 crystal and fastrtps, in the first release of the commands.

### Frequently asked questions

- _How can we run the resulting binaries?_ Packaging the resulting binaries is
  outside of the scope of this document.
  The cross-compiled binaries will be
  available in the workspace in the cc-root (e.g.
  `generic_armhf-ubuntu_bionic-fastrtps-crystal`).
  We make the assumption that
  the user will copy the whole workspace to the remote device (including `src`)
  and run `rosdep` on the target platform, to install the dependencies before
  running the binaries.

### Maintenance and testing

Regarding maintenance, in order to support cross-compilation for a new platform
we would need to:

- Prepare and publish a base ROS 2 Docker image for the new platform, both
  fetching the dependencies using the ROS 2 source using `rosdep install`, and
  from the pre-built binary if available for that platform.
- Adapt the toolchain file for the new platform, if required.
  For that the toolchains in [ruslo/polly](https://github.com/ruslo/polly) might be useful.

> - Add a new `cross-compile*` colcon mixin to for the new platform and perform

Reviewer 6: There's a subject missing.

"Add a new `cross-compile*` colcon mixin to ______ for the new platform..."
